pkg/kubelet/status/manager.go

```diff
@@ -205,16 +175,37 @@ func (m *manager) SetContainerReadiness(pod *api.Pod, containerID kubecontainer.
 		return
 	}
 
-	if status.ContainerStatuses[containerIndex].Ready == ready {
+	if oldStatus.status.ContainerStatuses[containerIndex].Ready == ready {
 		glog.V(4).Infof("Container readiness unchanged (%v): %q - %q", ready,
 			format.Pod(pod), containerID.String())
 		return
 	}
 
 	// Make sure we're not updating the cached version.
-	status.ContainerStatuses = make([]api.ContainerStatus, len(status.ContainerStatuses))
-	copy(status.ContainerStatuses, oldStatus.status.ContainerStatuses)
+	clone, err := api.Scheme.DeepCopy(&oldStatus.status)
+	if err != nil {
+		glog.Errorf("Failed to clone status %+v: %v", oldStatus.status, err)
+		return
+	}
+	status := *clone.(*api.PodStatus)
 	status.ContainerStatuses[containerIndex].Ready = ready
+
+	// Update pod condition.
+	readyConditionIndex := -1
+	for i, condition := range status.Conditions {
+		if condition.Type == api.PodReady {
+			readyConditionIndex = i
+			break
+		}
+	}
+	readyCondition := GeneratePodReadyCondition(&pod.Spec, status.ContainerStatuses, status.Phase)
+	if readyConditionIndex != -1 {
+		status.Conditions[readyConditionIndex] = readyCondition
+	} else {
+		glog.Warningf("PodStatus missing PodReady condition: %+v", status)
```

Comment:
After reading through [kubelet.generatePodStatus](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L3108), I don't think this should ever happen. I think in all the cases where there wouldn't be a ready condition, there also wouldn't be container statuses, so this function would abort earlier.

I can change it to a higher info level if you prefer though.